File: frontend/src/types/post.ts
// public: 전체공개 / subscribers: 구독자공개 / private: 나만 보기
export type Visibility = 'public' | 'subscribers' | 'private'

// 백엔드 PostRead 스키마와 같은 모양
export interface Post {
  id: number
  title: string
  content: string
  cover_image: string | null // 커버(대표) 이미지 URL, 없으면 null
  tags: string[] // 태그 (없으면 빈 배열)
  series: string | null // 연재 이름 (같은 이름끼리 한 시리즈), 없으면 null
  owner_id: number | null
  author_name: string | null
  author_handle: string | null
  visibility: Visibility
  created_at: string
  updated_at: string
}

// 연재 네비 (서버 SeriesNav와 동일 모양). 연재가 아니면 API가 null을 준다.
export interface SeriesItem {
  id: number
  title: string
  created_at: string
}

export interface SeriesNav {
  series: string
  total: number // 이 연재에서 '내가 볼 수 있는' 글 수
  index: number // 이 글이 몇 번째인지 (1부터)
  items: SeriesItem[]
  prev: SeriesItem | null // 이전 편(더 오래된 글)
  next: SeriesItem | null // 다음 편(더 최신 글)
}

// 목록용 — 본문 전체 대신 발췌+읽기시간 (서버 PostSummary와 동일 모양)
export interface PostSummary {
  id: number
  title: string
  excerpt: string
  reading_minutes: number
  cover_image: string | null
  tags: string[]
  series: string | null
  owner_id: number | null
  // 글쓴이 표시용 (2026-08-27). handle 은 **공개 블로그를 가질 수 있는 역할일 때만**
  // 서버가 채운다 — 차단된 사람의 `/@handle` 은 빈 목록이라 링크를 그리면 빈 페이지로
  // 보내게 된다. 그래서 화면은 `author_handle` 이 있을 때만 링크로, 없으면 이름만 그린다.
  author_name: string | null
  author_handle: string | null
  visibility: Visibility
  created_at: string
  updated_at: string
}

File: backend/app/models/post.py
from datetime import datetime
from sqlalchemy import String, Text, DateTime, ForeignKey, func, Index
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.orm import Mapped, mapped_column
from app.core.database import Base


class Post(Base):
    __tablename__ = "posts"
    __table_args__ = (
        # 태그 필터(tags @> ARRAY[...])가 전체스캔 대신 인덱스를 타도록 GIN 인덱스
        Index("ix_posts_tags", "tags", postgresql_using="gin"),
        # 검색(ILIKE '%…%')용 trigram GIN 인덱스. 한국어는 to_tsvector가 형태소를
        # 몰라 풀텍스트가 안 먹으므로 pg_trgm으로 부분일치를 인덱스 스캔한다.
        # 주의: trigram이라 2글자 이하 검색어는 인덱스를 못 타고 순차 스캔이 된다.
        Index(
            "ix_posts_title_trgm",
            "title",
            postgresql_using="gin",
            postgresql_ops={"title": "gin_trgm_ops"},
        ),
        Index(
            "ix_posts_content_trgm",
            "content",
            postgresql_using="gin",
            postgresql_ops={"content": "gin_trgm_ops"},
        ),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    title: Mapped[str] = mapped_column(String(200))
    content: Mapped[str] = mapped_column(Text)
    # 커버(대표) 이미지 URL — 선택. /api/upload 로 올린 이미지 URL을 저장. 없으면 None
    cover_image: Mapped[str | None] = mapped_column(String(500), nullable=True)
    # 태그(다중) — Postgres 텍스트 배열로 저장. 기본 빈 배열. 태그로 글 필터에 사용
    tags: Mapped[list[str]] = mapped_column(ARRAY(String), server_default="{}")
    # 연재 이름 — 같은 값을 가진 글이 한 시리즈가 되고, 순서는 created_at.
    # 제목의 '#7' 같은 번호를 파싱하지 않는 이유: 제목을 고치면 순서가 깨지고,
    # 번호를 매기려면 글마다 손으로 붙여야 한다. 이름만 같으면 되게 두는 게 단순하다.
    series: Mapped[str | None] = mapped_column(String(100), nullable=True, index=True)
    # 작성자. 기존(로그인 전) 글은 owner 없음 → nullable
    owner_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id"), nullable=True, index=True
    )
    # 공개범위: 'public'(전체공개) | 'subscribers'(구독자공개) | 'private'(나만 보기)
    # 'subscribers'가 11자라 varchar(20)으로 둠. 기존 글은 public
    visibility: Mapped[str] = mapped_column(String(20), server_default="public")
    # server_default=func.now(): DB가 INSERT 시 현재 시각을 채움
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    # onupdate: UPDATE 될 때마다 시각 자동 갱신
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )
